lib/scholar/neighbors/k_nearest_neighbors.ex

```diff
@@ -0,0 +1,401 @@
+      {#Nx.Tensor<
+        f32[2][4]
+        [
+          [0.3162279427051544, 0.7071065902709961, 1.5811389684677124, 2.469817876815796],
+          [0.10000002384185791, 1.0049875974655151, 2.193171262741089, 3.132091760635376]
+        ]
+      >,
+      #Nx.Tensor<
+        s64[2][4]
+        [
+          [1, 3, 2, 0],
+          [0, 2, 1, 3]
+        ]
+      >}
+  """
+  defn k_neighbors(
+         %__MODULE__{
+           data: data,
+           default_num_neighbors: default_num_neighbors,
+           p: p,
+           metric: metric
+         } = _model,
+         x
+       ) do
+    {num_samples, num_features} = Nx.shape(data)
+    {num_samples_x, _num_features} = Nx.shape(x)
+
+    dist =
+      case metric do
+        :minkowski ->
+          Scholar.Metrics.Distance.minkowski(
+            Nx.new_axis(data, 0) |> Nx.broadcast({num_samples_x, num_samples, num_features}),
+            Nx.new_axis(x, 1) |> Nx.broadcast({num_samples_x, num_samples, num_features}),
+            axes: [-1],
+            p: p
+          )
+
+        :cosine ->
+          Scholar.Metrics.Distance.cosine(
+            Nx.new_axis(data, 0) |> Nx.broadcast({num_samples_x, num_samples, num_features}),
+            Nx.new_axis(x, 1) |> Nx.broadcast({num_samples_x, num_samples, num_features}),
+            axes: [-1]
+          )
+      end
+
+    {val, ind} = Nx.top_k(-dist, k: default_num_neighbors)
+    {-val, ind}
+  end
+
+  defnp mode_weighted(tensor, weights, opts \\ []) do
+    axis = opts[:axis]
+    tensor_size = Nx.size(tensor)
+
+    cond do
+      tensor_size == 1 ->
+        Nx.squeeze(tensor, axes: [axis])
+
+      true ->
+        weighted_mode_general(tensor, weights, axis: axis)
+    end
+  end
+
+  defnp weighted_mode_general(tensor, weights, opts) do
+    axis = opts[:axis]
+
+    {tensor, weights, axis} =
+      if axis == 0,
+        do: {Nx.transpose(tensor), Nx.transpose(weights), 1},
+        else: {tensor, weights, axis}
+
+    {num_samples, num_features} = tensor_shape = Nx.shape(tensor)
+
+    indices = Nx.argsort(tensor, axis: axis)
+
+    sorted = Nx.sort(tensor, axis: axis)
+
+    size_to_broadcast = {num_samples, 1}
+
+    group_indices =
+      Nx.concatenate(
+        [
+          Nx.broadcast(0, size_to_broadcast),
+          Nx.not_equal(
+            Nx.slice_along_axis(sorted, 0, Nx.axis_size(sorted, axis) - 1, axis: axis),
+            Nx.slice_along_axis(sorted, 1, Nx.axis_size(sorted, axis) - 1, axis: axis)
+          )
+        ],
+        axis: axis
+      )
+      |> Nx.cumulative_sum(axis: axis)
+
+    num_elements = Nx.size(tensor_shape)
+
+    counting_indices =
+      [
+        Nx.shape(group_indices)
+        |> Nx.iota(axis: 0)
+        |> Nx.reshape({num_elements, 1}),
+        Nx.reshape(group_indices, {num_elements, 1})
+      ]
+      |> Nx.concatenate(axis: 1)
+
+    to_add = Nx.broadcast(1, {num_elements}) * (weights |> Nx.reshape({:auto}))
```

Comment:
Isn't this the same as `Nx.flatten(weights)`?

I mean, if the shapes differ anyway, the multiplication won't work as there is no `1`-sized axis in neither tensor for broadcasting, so in the end this is just an element-wise multiplication by 1

Reply:
Ofc you're right, I will fix that